modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/PaginationBarTag.java

```diff
@@ -288,50 +301,134 @@ protected int processStartTag() throws Exception {
 
 			buttonTag.setCssClass("page-link");
 			buttonTag.setDisplayType("unstyled");
-			buttonTag.setLabel(String.valueOf(i));
+			buttonTag.setLabel(String.valueOf(firstItem));
 
 			buttonTag.doTag(pageContext);
 
-			jspWriter.write("</li>");
-		}
+			int leftIndex = activeIndex - _ellipsisBuffer;
 
-		jspWriter.write("<li class=\"dropdown page-item\">");
+			int[] leftItems = Arrays.copyOfRange(
+				totalItems, 1, Math.max(leftIndex, 1));
 
-		buttonTag = new ButtonTag();
+			if (leftItems.length > 1) {
+				jspWriter.write("<li class=\"dropdown page-item\">");
 
-		buttonTag.setCssClass("dropdown-toggle page-link");
-		buttonTag.setDisplayType("unstyled");
-		buttonTag.setLabel("...");
+				buttonTag = new ButtonTag();
 
-		buttonTag.doTag(pageContext);
+				buttonTag.setCssClass("dropdown-toggle page-link");
+				buttonTag.setDisplayType("unstyled");
+				buttonTag.setLabel("...");
 
-		jspWriter.write("</li>");
+				buttonTag.doTag(pageContext);
 
-		for (int i = ellipsisCount + 1; i < (totalPages + 1); i++) {
-			jspWriter.write("<li class=\"page-item");
+				jspWriter.write("</li>");
+			}
 
-			if (i == _activePage) {
-				jspWriter.write(" active");
+			int lastIndex = totalItems.length - 1;
+
+			int[] centerItems = Arrays.copyOfRange(
+				totalItems, Math.max(activeIndex - _ellipsisBuffer, 1),
+				Math.min(activeIndex + _ellipsisBuffer + 1, lastIndex));
+
+			for (int itemToRender : centerItems) {
+				jspWriter.write("<li class=\"page-item");
```

Comment:
Maybe it is a good strategy to create some auxiliary methods to create the page items to add more meaning, clarify the pagination bar creation strategy and avoid duplicating the code.

You can create a private addPageItem method with the following code
```
jspWriter.write("<li class=\"page-item");

if (itemToRender == _activePage) {
    jspWriter.write(" active");
}

if ((_disabledPages != null) &&
	_disabledPages.contains(Integer.valueOf(itemToRender))) {

    jspWriter.write(" disabled");
}

jspWriter.write("\">");

buttonTag = new ButtonTag();

buttonTag.setCssClass("page-link");
buttonTag.setDisplayType("unstyled");
buttonTag.setLabel(String.valueOf(itemToRender));

buttonTag.doTag(pageContext);
```

Another good candidate is to also create a method for addDropdownPageItem.

Even you can extract methods create the pagination with dropdown and another to create the pagination without dropdown, although Brian usually doesn't like to extract methods that are only called once

Reply:
@javierdearcos as I mentioned in the previous PR, up to now we decided to keep all the "rendering" part in the `processStartTag` method. If you have a look at [ManagementToolbarTag](https://github.com/liferay/liferay-portal/blob/6fa77fbbc8f10e3fb7418bb2b75a7a62de2143de/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/ManagementToolbarTag.java#L682-L1113), you'll see that it's `processStartTag` method is really "big", when I initially submitted the pull request I had added separate method to avoid repeating myself and avoid having such a big method - so I'd be careful with this.

Reply:
I suggested the same but I think it's a "taglib pattern" to have everything in the same method.

Reply:
:+1: Thanks

Reply:
Ohh, ok. Didn't know about it. Good to know. So let's leave it as it is. As I said, was just recommendations and nothing mandatory or needed

Reply:
> I suggested the same but I think it's a "taglib pattern" to have everything in the same method.

I don't think there's such a thing... at least not that I'm aware of. I don't see why we wouldn't want to split the logic in `processStartTag` with some helpers to make it more manageable 🤷

Reply:
Julien told me that they tried to extract methods in another taglib. But we can always try :)

Reply:
Julien told me that they tried to extract methods in another taglib. But we can try :D

Reply:
Can't remember what happened (and if I was around), but you can check one of our oldest tags like [IcontTag](https://github.com/liferay-frontend/liferay-portal/blob/master/util-taglib/src/com/liferay/taglib/aui/IconTag.java#L91) for this.

Reply:
Thanks @jbalsas that clarified my doubts. 

I've re-submitted a pull request [here](https://github.com/liferay-frontend/liferay-portal/pull/1525) , Brian [told me](https://github.com/brianchandotcom/liferay-portal/pull/108036#issuecomment-931347386) I could bypass CI, but I want to make sure we're all "happy" with the change.

/cc @javierdearcos